addon-api/content-script/contextmenu.js

```diff
@@ -2,34 +2,94 @@ let initialized = false;
 let hasDynamicContextMenu = false;
 let contextMenus = [];
 
-const onReactContextMenu = function (e) {
+const findParentWithProp = (reactInternalInstance, prop) => {
+  if (!reactInternalInstance) return null;
+  while (
+    !reactInternalInstance.stateNode?.props ||
+    !Object.prototype.hasOwnProperty.call(reactInternalInstance.stateNode.props, prop)
+  ) {
+    if (!reactInternalInstance.return) return null;
+    reactInternalInstance = reactInternalInstance.return;
+  }
+  return reactInternalInstance.stateNode;
+};
+const findMonitor = (reactInternalInstance) => findParentWithProp(reactInternalInstance, "opcode");
+const findSpriteSelectorItem = (reactInternalInstance) => findParentWithProp(reactInternalInstance, "dragType");
+
+const setFocus = (item) => {
+  item.setAttribute("data-highlighted", "");
+  item.tabIndex = 0;
+  item.focus();
+};
+
+const removeFocus = (item) => {
+  item.removeAttribute("data-highlighted");
+  item.tabIndex = -1;
+};
+
+const menuArrowKeyListener = (menu) => (e) => {
+  if (e.target !== menu) {
+    // Target is one of the items, not the menu
+    return;
+  }
+  const moveFocusTo = (newFocusedItem) => {
+    e.stopPropagation();
+    setFocus(newFocusedItem);
+  };
+  if (["Home", "PageUp", "ArrowDown"].includes(e.key)) moveFocusTo(menu.firstElementChild);
+  else if (["End", "PageDown", "ArrowUp"].includes(e.key)) moveFocusTo(menu.lastElementChild);
+};
+
+const itemArrowKeyListener = (menu, item) => (e) => {
+  const moveFocusTo = (newFocusedItem) => {
+    e.stopPropagation();
+    removeFocus(item);
+    setFocus(newFocusedItem);
+  };
+  if (e.key === "ArrowDown" && item.nextElementSibling) moveFocusTo(item.nextElementSibling);
+  else if (e.key === "ArrowUp" && item.previousElementSibling) moveFocusTo(item.previousElementSibling);
+  else if (["Home", "PageUp"].includes(e.key)) moveFocusTo(menu.firstElementChild);
+  else if (["End", "PageDown"].includes(e.key)) moveFocusTo(menu.lastElementChild);
+};
+
+const onReactContextMenu = async function (e) {
+  // This function expects "this" to be an addon.tab instance.
+
   if (!e.target) return;
-  const ctxTarget = e.target.closest(".react-contextmenu-wrapper");
+  const ctxTarget = e.target.closest(".react-contextmenu-wrapper, [data-state]");
   if (!ctxTarget) return;
-  let ctxMenu = ctxTarget.querySelector("nav.react-contextmenu");
+
+  const isReact18 = !ctxTarget.classList.contains("react-contextmenu-wrapper");
+  let ctxMenu;
+  if (isReact18) ctxMenu = await this.waitForElement("[data-radix-menu-content]");
```

Comment:
Have you tested whether this could be replaced with `queueMicrotask(() => document.querySelector())` instead?

Reply:
I can't test this right now because I'm not able to build the React 18 branch for some reason.